// @remotion/studio/dist/components/Timeline/TimelineSequenceName.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TimelineSequenceName = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const TimelineSelection_1 = require("./TimelineSelection");
const MAX_DISPLAY_NAME_LENGTH = 1000;
const getTruncatedDisplayName = (displayName) => {
    if (displayName.length > MAX_DISPLAY_NAME_LENGTH) {
        return displayName.slice(0, MAX_DISPLAY_NAME_LENGTH) + '...';
    }
    return displayName;
};
const TimelineSequenceName = ({ sequence, selected, containsSelection }) => {
    const style = (0, react_1.useMemo)(() => {
        return {
            alignItems: 'center',
            alignSelf: 'stretch',
            ...(0, TimelineSelection_1.getTimelineSelectedLabelStyle)(selected, false),
            display: 'inline-flex',
            fontSize: 12,
            whiteSpace: 'nowrap',
            textOverflow: 'ellipsis',
            color: (0, TimelineSelection_1.getTimelineColor)(selected, false),
            userSelect: 'none',
            WebkitUserSelect: 'none',
            textDecoration: 'none',
            boxShadow: containsSelection && !selected
                ? `inset 0 0 0 2px ${TimelineSelection_1.TIMELINE_SELECTED_LABEL_BACKGROUND}`
                : undefined,
        };
    }, [selected, containsSelection]);
    const text = getTruncatedDisplayName(sequence.displayName) || '<Sequence>';
    return (jsx_runtime_1.jsx("div", { title: text, style: style, children: text }));
};
exports.TimelineSequenceName = TimelineSequenceName;
