// @remotion/studio/dist/components/import-assets.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.insertComponent = exports.insertExistingAssets = exports.insertRemoteAudio = exports.importRemoteAsset = exports.importAssets = exports.pickFilesToImport = exports.getAssetElementFromPath = exports.getAssetElement = void 0;
const studio_shared_1 = require("@remotion/studio-shared");
const get_static_files_1 = require("../api/get-static-files");
const write_static_file_1 = require("../api/write-static-file");
const install_required_package_1 = require("../helpers/install-required-package");
const call_api_1 = require("./call-api");
const NotificationCenter_1 = require("./Notifications/NotificationCenter");
const getAssetElement = ({ fileType, src, }) => {
    if (fileType.type === 'png' ||
        fileType.type === 'jpeg' ||
        fileType.type === 'webp' ||
        fileType.type === 'bmp') {
        return {
            type: 'asset',
            assetType: 'image',
            src,
            srcType: 'static',
            dimensions: fileType.dimensions,
        };
    }
    if (fileType.type === 'gif') {
        return {
            type: 'asset',
            assetType: 'gif',
            src,
            srcType: 'static',
            dimensions: fileType.dimensions,
        };
    }
    if (fileType.type === 'riff' ||
        fileType.type === 'webm' ||
        fileType.type === 'iso-base-media' ||
        fileType.type === 'transport-stream') {
        return {
            type: 'asset',
            assetType: 'video',
            src,
            srcType: 'static',
            dimensions: null,
        };
    }
    if (fileType.type === 'wav' ||
        fileType.type === 'mp3' ||
        fileType.type === 'aac' ||
        fileType.type === 'flac') {
        return {
            type: 'asset',
            assetType: 'audio',
            src,
            srcType: 'static',
            dimensions: null,
        };
    }
    return null;
};
exports.getAssetElement = getAssetElement;
const getAssetElementFromPath = (assetPath) => {
    var _a;
    if (!assetPath || assetPath.includes('\\')) {
        return null;
    }
    const extension = (_a = assetPath.split('.').pop()) === null || _a === void 0 ? void 0 : _a.toLowerCase();
    if (!extension || extension === assetPath.toLowerCase()) {
        return null;
    }
    if (['png', 'jpg', 'jpeg', 'webp', 'bmp'].includes(extension)) {
        return {
            type: 'asset',
            assetType: 'image',
            src: assetPath,
            srcType: 'static',
            dimensions: null,
        };
    }
    if (extension === 'gif') {
        return {
            type: 'asset',
            assetType: 'gif',
            src: assetPath,
            srcType: 'static',
            dimensions: null,
        };
    }
    if (['mp4', 'm4v', 'mov', 'avi', 'webm', 'ts', 'm2ts'].includes(extension)) {
        return {
            type: 'asset',
            assetType: 'video',
            src: assetPath,
            srcType: 'static',
            dimensions: null,
        };
    }
    if (['wav', 'mp3', 'aac', 'flac'].includes(extension)) {
        return {
            type: 'asset',
            assetType: 'audio',
            src: assetPath,
            srcType: 'static',
            dimensions: null,
        };
    }
    return null;
};
exports.getAssetElementFromPath = getAssetElementFromPath;
const getAssetLabel = (element) => {
    if (element.type !== 'asset') {
        throw new Error('Expected asset element');
    }
    if (element.assetType === 'image') {
        return '<Img>';
    }
    if (element.assetType === 'video') {
        return '<Video>';
    }
    if (element.assetType === 'gif') {
        return '<Gif>';
    }
    if (element.assetType === 'audio') {
        return '<Audio>';
    }
    throw new Error('Unsupported asset type');
};
const getComponentLabel = (component) => {
    return `<${component.componentName}>`;
};
const pickFilesToImport = () => {
    return new Promise((resolve) => {
        const input = document.createElement('input');
        input.type = 'file';
        input.multiple = true;
        input.style.display = 'none';
        let didResolve = false;
        const resolveOnce = (files) => {
            if (didResolve) {
                return;
            }
            didResolve = true;
            input.removeEventListener('change', onChange);
            input.removeEventListener('cancel', onCancel);
            input.remove();
            resolve(files);
        };
        const onChange = () => {
            var _a;
            return resolveOnce(Array.from((_a = input.files) !== null && _a !== void 0 ? _a : []));
        };
        const onCancel = () => resolveOnce([]);
        input.addEventListener('change', onChange);
        input.addEventListener('cancel', onCancel);
        document.body.appendChild(input);
        input.click();
    });
};
exports.pickFilesToImport = pickFilesToImport;
const notifyInsertedAssets = (insertedLabels) => {
    if (insertedLabels.length === 1) {
        (0, NotificationCenter_1.showNotification)(`Added ${insertedLabels[0]} to source file`, 2000);
    }
    else if (insertedLabels.length > 1) {
        (0, NotificationCenter_1.showNotification)(`Added ${insertedLabels.length} assets to source file`, 2000);
    }
};
const notifyUnsupportedFiles = (unsupportedFiles) => {
    if (unsupportedFiles.length === 1) {
        (0, NotificationCenter_1.showNotification)(`Cannot add ${unsupportedFiles[0]}: Unsupported file type`, 3000);
    }
    else if (unsupportedFiles.length > 1) {
        (0, NotificationCenter_1.showNotification)(`Skipped ${unsupportedFiles.length} unsupported files`, 3000);
    }
};
const insertAssetElement = async ({ compositionFile, compositionId, element, }) => {
    const requiredPackage = (0, studio_shared_1.getRequiredPackageForInsertableElement)(element);
    await (0, install_required_package_1.installRequiredPackages)(requiredPackage ? [requiredPackage] : []);
    const result = await (0, call_api_1.callApi)('/api/insert-jsx-element', {
        compositionFile,
        compositionId,
        element,
    });
    if (!result.success) {
        (0, NotificationCenter_1.showNotification)(result.reason, 4000);
        return false;
    }
    return true;
};
const downloadRemoteAsset = async (url) => {
    return (0, call_api_1.callApi)('/api/download-remote-asset', { url });
};
const importAssets = async ({ compositionFile, compositionId, files, }) => {
    if (files.length === 0) {
        return;
    }
    const staticFiles = (0, get_static_files_1.getStaticFiles)();
    const differentExistingFile = files.find((file) => {
        return staticFiles.some((staticFile) => staticFile.name === file.name && staticFile.sizeInBytes !== file.size);
    });
    if (differentExistingFile) {
        (0, NotificationCenter_1.showNotification)(`File with name ${differentExistingFile.name} already exists and is different`, 4000);
        return;
    }
    const insertedLabels = [];
    const addedStaticFiles = [];
    const unsupportedFiles = [];
    const notifyAddedStaticFiles = () => {
        if (addedStaticFiles.length === 1) {
            (0, NotificationCenter_1.showNotification)(`Created ${addedStaticFiles[0]} in public folder`, 3000);
        }
        else if (addedStaticFiles.length > 1) {
            (0, NotificationCenter_1.showNotification)(`Added ${addedStaticFiles.length} files to public folder`, 3000);
        }
        addedStaticFiles.length = 0;
    };
    try {
        for (const file of files) {
            const contents = await file.arrayBuffer();
            const fileType = (0, studio_shared_1.detectFileType)(new Uint8Array(contents));
            const element = (0, exports.getAssetElement)({
                fileType,
                src: file.name,
            });
            if (element === null) {
                unsupportedFiles.push(file.name);
                continue;
            }
            const alreadyExists = staticFiles.some((staticFile) => staticFile.name === file.name && staticFile.sizeInBytes === file.size);
            if (!alreadyExists) {
                await (0, write_static_file_1.writeStaticFile)({
                    contents,
                    filePath: file.name,
                });
                addedStaticFiles.push(file.name);
            }
            const inserted = await insertAssetElement({
                compositionFile,
                compositionId,
                element,
            });
            if (!inserted) {
                notifyAddedStaticFiles();
                return;
            }
            insertedLabels.push(getAssetLabel(element));
        }
        notifyAddedStaticFiles();
        notifyInsertedAssets(insertedLabels);
        notifyUnsupportedFiles(unsupportedFiles);
    }
    catch (error) {
        (0, NotificationCenter_1.showNotification)(`Could not add asset: ${error instanceof Error ? error.message : String(error)}`, 4000);
    }
};
exports.importAssets = importAssets;
const importRemoteAsset = async ({ compositionFile, compositionId, url, }) => {
    try {
        const { assetPath, created, element } = await downloadRemoteAsset(url);
        if (created) {
            (0, NotificationCenter_1.showNotification)(`Created ${assetPath} in public folder`, 3000);
        }
        const inserted = await insertAssetElement({
            compositionFile,
            compositionId,
            element,
        });
        if (!inserted) {
            return;
        }
        notifyInsertedAssets([getAssetLabel(element)]);
    }
    catch (error) {
        (0, NotificationCenter_1.showNotification)(`Could not add remote asset: ${error instanceof Error ? error.message : String(error)}`, 4000);
    }
};
exports.importRemoteAsset = importRemoteAsset;
const insertRemoteAudio = async ({ compositionFile, compositionId, url, }) => {
    if (!(0, studio_shared_1.isUrl)(url)) {
        (0, NotificationCenter_1.showNotification)('Cannot add sound effect: Unsupported URL', 3000);
        return;
    }
    const element = {
        type: 'asset',
        assetType: 'audio',
        src: url,
        srcType: 'remote',
        dimensions: null,
    };
    try {
        const inserted = await insertAssetElement({
            compositionFile,
            compositionId,
            element,
        });
        if (!inserted) {
            return;
        }
        notifyInsertedAssets([getAssetLabel(element)]);
    }
    catch (error) {
        (0, NotificationCenter_1.showNotification)(`Could not add sound effect: ${error instanceof Error ? error.message : String(error)}`, 4000);
    }
};
exports.insertRemoteAudio = insertRemoteAudio;
const insertExistingAssets = async ({ assetPaths, compositionFile, compositionId, }) => {
    if (assetPaths.length === 0) {
        return;
    }
    const insertedLabels = [];
    const unsupportedFiles = [];
    try {
        for (const assetPath of assetPaths) {
            const element = (0, exports.getAssetElementFromPath)(assetPath);
            if (element === null) {
                unsupportedFiles.push(assetPath);
                continue;
            }
            const inserted = await insertAssetElement({
                compositionFile,
                compositionId,
                element,
            });
            if (!inserted) {
                return;
            }
            insertedLabels.push(getAssetLabel(element));
        }
        notifyInsertedAssets(insertedLabels);
        notifyUnsupportedFiles(unsupportedFiles);
    }
    catch (error) {
        (0, NotificationCenter_1.showNotification)(`Could not add asset: ${error instanceof Error ? error.message : String(error)}`, 4000);
    }
};
exports.insertExistingAssets = insertExistingAssets;
const insertComponent = async ({ component, compositionFile, compositionId, }) => {
    try {
        const inserted = await insertAssetElement({
            compositionFile,
            compositionId,
            element: {
                type: 'component',
                componentName: component.componentName,
                importName: component.importName,
                importPath: component.importPath,
                props: component.props,
            },
        });
        if (!inserted) {
            return;
        }
        (0, NotificationCenter_1.showNotification)(`Added ${getComponentLabel(component)} to source file`, 2000);
    }
    catch (error) {
        (0, NotificationCenter_1.showNotification)(`Could not add component: ${error instanceof Error ? error.message : String(error)}`, 4000);
    }
};
exports.insertComponent = insertComponent;
