// @remotion/studio/dist/components/SelectedOutlineOverlay.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.SelectedOutlineOverlay = exports.compensateTranslateForTransformOrigin = exports.getSelectedOutlineRotationDragChanges = exports.getSelectedOutlineRotationDragValues = exports.getSelectedOutlineRotationDragStates = exports.getSelectedOutlineScaleDragChanges = exports.getSelectedOutlineScaleDragValues = exports.getSelectedOutlineScaleDragStates = exports.getSelectedOutlineScaleEdgeInfo = exports.getSelectedOutlineDragChanges = exports.applySelectedOutlineDragAxisLock = exports.getSelectedOutlineDragValues = exports.getSequencesWithSelectableOutlines = exports.getSelectedEffectFieldsBySequenceKey = exports.getOutlineSelectionInteraction = exports.getSelectedSequenceKeys = exports.getSelectedOutlineRotationCornerInfo = exports.getSelectedOutlineRotationDeltaDegrees = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = __importStar(require("react"));
const remotion_1 = require("remotion");
const no_react_1 = require("remotion/no-react");
const calculate_timeline_1 = require("../helpers/calculate-timeline");
const client_id_1 = require("../helpers/client-id");
const colors_1 = require("../helpers/colors");
const format_file_location_1 = require("../helpers/format-file-location");
const get_box_quads_ponyfill_1 = require("../helpers/get-box-quads-ponyfill");
const open_in_editor_1 = require("../helpers/open-in-editor");
const editor_outlines_1 = require("../state/editor-outlines");
const scale_lock_1 = require("../state/scale-lock");
const ContextMenu_1 = require("./ContextMenu");
const effect_drag_and_drop_1 = require("./effect-drag-and-drop");
const ForceSpecificCursor_1 = require("./ForceSpecificCursor");
const NotificationCenter_1 = require("./Notifications/NotificationCenter");
const selected_outline_geometry_1 = require("./selected-outline-geometry");
const selected_outline_uv_1 = require("./selected-outline-uv");
const SelectedOutlineUvControls_1 = require("./SelectedOutlineUvControls");
const call_add_keyframe_1 = require("./Timeline/call-add-keyframe");
const parse_keyframe_field_from_node_path_1 = require("./Timeline/parse-keyframe-field-from-node-path");
const save_sequence_prop_1 = require("./Timeline/save-sequence-prop");
const timeline_rotation_utils_1 = require("./Timeline/timeline-rotation-utils");
const timeline_translate_utils_1 = require("./Timeline/timeline-translate-utils");
const TimelineScaleField_1 = require("./Timeline/TimelineScaleField");
const TimelineSelection_1 = require("./Timeline/TimelineSelection");
const get_stack_1 = require("./Timeline/TimelineStack/get-stack");
const transform_origin_utils_1 = require("./Timeline/transform-origin-utils");
const translateFieldKey = 'style.translate';
const scaleFieldKey = 'style.scale';
const rotateFieldKey = 'style.rotate';
const transformOriginFieldKey = 'style.transformOrigin';
const outlineContainer = {
    position: 'absolute',
    inset: 0,
    pointerEvents: 'none',
    overflow: 'visible',
};
const emptyContextMenuValues = [];
const pointToString = (point) => `${point.x},${point.y}`;
const midpoint = (from, to) => {
    return (0, selected_outline_geometry_1.mixPoint)(from, to, 0.5);
};
const getOutlineCenter = (points) => {
    const [tl, tr, br, bl] = points;
    return {
        x: (tl.x + tr.x + br.x + bl.x) / 4,
        y: (tl.y + tr.y + br.y + bl.y) / 4,
    };
};
const dot = (left, right) => {
    return left.x * right.x + left.y * right.y;
};
const vectorLength = (vector) => {
    return Math.hypot(vector.x, vector.y);
};
const vectorBetween = (from, to) => {
    return { x: to.x - from.x, y: to.y - from.y };
};
const getAngleDegrees = (from, to) => {
    return Math.atan2(to.y - from.y, to.x - from.x) * (180 / Math.PI);
};
const getSelectedOutlineRotationDeltaDegrees = ({ from, to, }) => {
    return ((((to - from) % 360) + 540) % 360) - 180;
};
exports.getSelectedOutlineRotationDeltaDegrees = getSelectedOutlineRotationDeltaDegrees;
const normalizeRotationCursorDegrees = (rotation) => {
    const normalizedRotation = ((rotation % 360) + 360) % 360;
    return Number(normalizedRotation.toFixed(3));
};
const getRotationCursor = (rotation) => {
    const normalizedRotation = normalizeRotationCursorDegrees(rotation);
    const transform = normalizedRotation === 0
        ? ''
        : `<g transform="rotate(${normalizedRotation} 32 32)">`;
    const transformEnd = normalizedRotation === 0 ? '' : '</g>';
    const svg = `<svg width="24" height="24" viewBox="0 0 64 64" fill="none" xmlns="http://www.w3.org/2000/svg">${transform}<g transform="scale(0.876712329)" filter="url(#filter0_d_1_14)"><path d="M10.9111 17.7687C10.3413 18.3701 10.367 19.3195 10.9684 19.8893L20.7687 29.1738C21.3701 29.7436 22.3195 29.7179 22.8893 29.1165C23.459 28.5151 23.4334 27.5657 22.832 26.996L14.1206 18.7431L22.3735 10.0316C22.9432 9.43022 22.9176 8.48082 22.3162 7.91107C21.7148 7.34133 20.7654 7.36699 20.1956 7.96839L10.9111 17.7687ZM49.3923 58.3118C49.9509 58.9235 50.8996 58.9667 51.5114 58.4081L61.481 49.3055C62.0927 48.7469 62.1359 47.7981 61.5773 47.1863C61.0187 46.5745 60.0699 46.5314 59.4581 47.09L50.5963 55.1812L42.5051 46.3194C41.9465 45.7076 40.9977 45.6645 40.386 46.2231C39.7742 46.7817 39.7311 47.7304 40.2896 48.3422L49.3923 58.3118ZM12.6747 18.7431L13 19.8893C22.1283 19.6426 30.7584 21.4283 37.8564 26.6927C44.8518 31.8809 49.734 39.8538 49 56H50.5963H51.5114C52.2774 39.1461 47.6482 30.2198 39.6436 24.2831C31.7416 18.4224 22.3717 17.2467 13 17.5L12.6747 18.7431Z" fill="black"/><path d="M19.1064 6.93652C20.2459 5.73379 22.1448 5.68278 23.3477 6.82227C24.5505 7.96181 24.6022 9.86076 23.4629 11.0635L18.7373 16.0508C26.3487 16.4239 33.9128 18.1651 40.5371 23.0781C44.7339 26.1907 48.0794 30.1189 50.2568 35.4834C51.9666 39.6958 52.9327 44.7395 53.0742 50.8867L58.4463 45.9824C59.6697 44.8654 61.5673 44.9514 62.6846 46.1748C63.8018 47.3985 63.7155 49.296 62.4922 50.4131L52.5225 59.5156C51.337 60.5979 49.5196 60.5507 48.3916 59.4346L48.2842 59.3232L39.1816 49.3535C38.0648 48.1301 38.1507 46.2324 39.374 45.1152C40.5975 43.9979 42.4961 44.0841 43.6133 45.3076L47.4756 49.5381C47.1908 44.7613 46.2876 40.9448 44.9482 37.8291C43.0666 33.4521 40.2851 30.3614 36.9629 27.8975C31.8259 24.0875 25.8071 22.1663 19.2891 21.5732L23.8633 25.9072C25.0662 27.0468 25.1178 28.9457 23.9785 30.1484C22.8746 31.3135 21.0576 31.3982 19.8516 30.3662L19.7373 30.2627L9.93652 20.9785C8.73384 19.839 8.6828 17.9401 9.82227 16.7373L19.1064 6.93652Z" stroke="white" stroke-width="3"/></g>${transformEnd}<defs><filter id="filter0_d_1_14" x="0" y="0" width="72.4696" height="72.3004" filterUnits="userSpaceOnUse" color-interpolation-filters="sRGB"><feFlood flood-opacity="0" result="BackgroundImageFix"/><feColorMatrix in="SourceAlpha" type="matrix" values="0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 127 0" result="hardAlpha"/><feOffset dy="3"/><feGaussianBlur stdDeviation="3.75"/><feComposite in2="hardAlpha" operator="out"/><feColorMatrix type="matrix" values="0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0.25 0"/><feBlend mode="normal" in2="BackgroundImageFix" result="effect1_dropShadow_1_14"/><feBlend mode="normal" in="SourceGraphic" in2="effect1_dropShadow_1_14" result="shape"/></filter></defs></svg>`;
    return `url("data:image/svg+xml,${encodeURIComponent(svg)}") 12 12, alias`;
};
const rotationCursorBaseDegrees = {
    'top-left': 270,
    'top-right': 0,
    'bottom-right': 90,
    'bottom-left': 180,
};
const getOutlineRotationDegrees = (points) => {
    const [tl, tr] = points;
    return getAngleDegrees(tl, tr);
};
const getRotationCursorDegrees = (points, corner) => normalizeRotationCursorDegrees(getOutlineRotationDegrees(points) + rotationCursorBaseDegrees[corner]);
const getSelectedOutlineRotationCornerInfo = (points, corner) => {
    const [tl, tr, br, bl] = points;
    const point = {
        'top-left': tl,
        'top-right': tr,
        'bottom-right': br,
        'bottom-left': bl,
    }[corner];
    const center = getOutlineCenter(points);
    const cursorDegrees = getRotationCursorDegrees(points, corner);
    return {
        center,
        cursor: getRotationCursor(cursorDegrees),
        cursorDegrees,
        point,
    };
};
exports.getSelectedOutlineRotationCornerInfo = getSelectedOutlineRotationCornerInfo;
const rectToPoints = (elementRect, containerRect) => {
    const left = elementRect.left - containerRect.left;
    const top = elementRect.top - containerRect.top;
    const right = elementRect.right - containerRect.left;
    const bottom = elementRect.bottom - containerRect.top;
    return [
        { x: left, y: top },
        { x: right, y: top },
        { x: right, y: bottom },
        { x: left, y: bottom },
    ];
};
const quadToPoints = (quad, containerRect) => {
    // `getBoxQuads`/the ponyfill returns the quad in viewport coordinates.
    // The overlay <svg> is unscaled (the canvas `scale()`/pan live on a sibling
    // container, not the svg), so 1 user unit == 1 px and we only need to move
    // the quad into the svg's local space by subtracting its viewport origin.
    // We deliberately do not pass `relativeTo` to the ponyfill: when the target
    // is not an ancestor of the element, the polyfill cannot resolve the
    // coordinate space and leaves the quad in viewport coordinates.
    return [
        { x: quad.p1.x - containerRect.left, y: quad.p1.y - containerRect.top },
        { x: quad.p2.x - containerRect.left, y: quad.p2.y - containerRect.top },
        { x: quad.p3.x - containerRect.left, y: quad.p3.y - containerRect.top },
        { x: quad.p4.x - containerRect.left, y: quad.p4.y - containerRect.top },
    ];
};
const getElementOutlinePoints = (element, containerRect) => {
    const elementRect = element.getBoundingClientRect();
    if (elementRect.width === 0 && elementRect.height === 0) {
        return null;
    }
    const quads = (0, get_box_quads_ponyfill_1.getBoxQuadsPonyfill)(element, {
        box: 'border',
    });
    const quad = quads === null || quads === void 0 ? void 0 : quads[0];
    if (!quad) {
        return rectToPoints(elementRect, containerRect);
    }
    return quadToPoints(quad, containerRect);
};
const getSelectedSequenceKeys = (selectedItems) => {
    return new Set(selectedItems
        .filter((item) => item.type === 'sequence')
        .map((item) => (0, TimelineSelection_1.getTimelineSequenceSelectionKey)(item.nodePathInfo)));
};
exports.getSelectedSequenceKeys = getSelectedSequenceKeys;
const getSequenceKeysContainingSelection = (selectedItems) => {
    return new Set(selectedItems.map((item) => (0, TimelineSelection_1.getTimelineSequenceSelectionKey)(item.nodePathInfo)));
};
const getOutlineSelectionInteraction = ({ shiftKey, metaKey, ctrlKey, }) => ({
    shiftKey,
    toggleKey: metaKey || ctrlKey,
});
exports.getOutlineSelectionInteraction = getOutlineSelectionInteraction;
const getSelectedEffectFieldsBySequenceKey = (selectedItems) => {
    var _a, _b;
    const selectedEffects = new Map();
    for (const item of selectedItems) {
        if (item.type !== 'sequence-effect' &&
            item.type !== 'sequence-effect-prop') {
            continue;
        }
        const sequenceKey = (0, TimelineSelection_1.getTimelineSequenceSelectionKey)(item.nodePathInfo);
        const effectsForSequence = (_a = selectedEffects.get(sequenceKey)) !== null && _a !== void 0 ? _a : new Map();
        const selectedFields = (_b = effectsForSequence.get(item.i)) !== null && _b !== void 0 ? _b : {
            allFields: false,
            fieldKeys: new Set(),
        };
        if (item.type === 'sequence-effect') {
            selectedFields.allFields = true;
        }
        else {
            selectedFields.fieldKeys.add(item.key);
        }
        effectsForSequence.set(item.i, selectedFields);
        selectedEffects.set(sequenceKey, effectsForSequence);
    }
    return selectedEffects;
};
exports.getSelectedEffectFieldsBySequenceKey = getSelectedEffectFieldsBySequenceKey;
const getSelectedTransformOriginInfo = (selectedItems) => {
    if (selectedItems.length !== 1) {
        return null;
    }
    const [selectedItem] = selectedItems;
    if (selectedItem.type === 'sequence-prop' &&
        selectedItem.key === transformOriginFieldKey) {
        return {
            sequenceKey: (0, TimelineSelection_1.getTimelineSequenceSelectionKey)(selectedItem.nodePathInfo),
            displayFrame: null,
        };
    }
    if (selectedItem.type !== 'keyframe') {
        return null;
    }
    const field = (0, parse_keyframe_field_from_node_path_1.parseKeyframeFieldFromNodePath)(selectedItem.nodePathInfo.auxiliaryKeys);
    if ((field === null || field === void 0 ? void 0 : field.type) !== 'sequence' ||
        field.fieldKey !== transformOriginFieldKey) {
        return null;
    }
    return {
        sequenceKey: (0, TimelineSelection_1.getTimelineSequenceSelectionKey)(selectedItem.nodePathInfo),
        displayFrame: selectedItem.frame,
    };
};
const getSequencesWithSelectableOutlines = ({ sequences, overrideIdsToNodePaths, }) => {
    return (0, calculate_timeline_1.calculateTimeline)({
        sequences: [...sequences],
        overrideIdsToNodePaths,
    })
        .filter((track) => {
        if (track.nodePathInfo === null) {
            return false;
        }
        return (track.sequence.showInTimeline &&
            track.nodePathInfo.auxiliaryKeys.length === 0);
    })
        .filter((track) => track.sequence.refForOutline !== null)
        .sort((a, b) => a.depth - b.depth)
        .map((track) => {
        if (track.nodePathInfo === null) {
            throw new Error('Expected selected outline to have a node path');
        }
        return {
            depth: track.depth,
            keyframeDisplayOffset: track.keyframeDisplayOffset,
            key: (0, TimelineSelection_1.getTimelineSequenceSelectionKey)(track.nodePathInfo),
            nodePathInfo: track.nodePathInfo,
            sequence: track.sequence,
        };
    });
};
exports.getSequencesWithSelectableOutlines = getSequencesWithSelectableOutlines;
const measureOutlines = (container, targets) => {
    const containerRect = container.getBoundingClientRect();
    const outlines = [];
    for (const target of targets) {
        const element = target.ref.current;
        if (element === null) {
            continue;
        }
        const points = getElementOutlinePoints(element, containerRect);
        if (points === null) {
            continue;
        }
        outlines.push({
            key: target.key,
            dimensions: element instanceof HTMLElement
                ? {
                    width: element.offsetWidth,
                    height: element.offsetHeight,
                }
                : null,
            points,
        });
    }
    return outlines;
};
const outlinesAreEqual = (a, b) => {
    var _a, _b, _c, _d;
    if (a.length !== b.length) {
        return false;
    }
    for (let i = 0; i < a.length; i++) {
        if (a[i].key !== b[i].key) {
            return false;
        }
        if (((_a = a[i].dimensions) === null || _a === void 0 ? void 0 : _a.width) !== ((_b = b[i].dimensions) === null || _b === void 0 ? void 0 : _b.width) ||
            ((_c = a[i].dimensions) === null || _c === void 0 ? void 0 : _c.height) !== ((_d = b[i].dimensions) === null || _d === void 0 ? void 0 : _d.height)) {
            return false;
        }
        for (let j = 0; j < a[i].points.length; j++) {
            if (Math.abs(a[i].points[j].x - b[i].points[j].x) > 0.01 ||
                Math.abs(a[i].points[j].y - b[i].points[j].y) > 0.01) {
                return false;
            }
        }
    }
    return true;
};
const getSelectedOutlineDragStates = ({ dragTargets, getDragOverrides, timelinePosition, }) => {
    return dragTargets.map((target) => {
        var _a;
        const dragOverrideValue = ((_a = getDragOverrides(target.nodePath)) !== null && _a !== void 0 ? _a : {})[translateFieldKey];
        const sourceFrame = timelinePosition - target.keyframeDisplayOffset;
        const effectiveValue = remotion_1.Internals.getEffectiveVisualModeValue({
            propStatus: target.propStatus,
            dragOverrideValue,
            defaultValue: target.fieldDefault,
            frame: sourceFrame,
            shouldResortToDefaultValueIfUndefined: true,
        });
        const [startX, startY] = (0, timeline_translate_utils_1.parseTranslate)(String(effectiveValue !== null && effectiveValue !== void 0 ? effectiveValue : '0px 0px'));
        return {
            defaultValue: target.fieldDefault !== undefined
                ? JSON.stringify(target.fieldDefault)
                : null,
            key: remotion_1.Internals.makeSequencePropsSubscriptionKey(target.nodePath),
            sourceFrame,
            startX,
            startY,
            target,
        };
    });
};
const getSelectedOutlineDragValues = ({ dragStates, deltaX, deltaY, }) => {
    return new Map(dragStates.map((dragState) => [
        dragState.key,
        (0, timeline_translate_utils_1.serializeTranslate)(dragState.startX + deltaX, dragState.startY + deltaY),
    ]));
};
exports.getSelectedOutlineDragValues = getSelectedOutlineDragValues;
const applySelectedOutlineDragAxisLock = ({ deltaX, deltaY, axisLocked, }) => {
    if (!axisLocked) {
        return { deltaX, deltaY };
    }
    if (Math.abs(deltaX) >= Math.abs(deltaY)) {
        return { deltaX, deltaY: 0 };
    }
    return { deltaX: 0, deltaY };
};
exports.applySelectedOutlineDragAxisLock = applySelectedOutlineDragAxisLock;
const getSelectedOutlineDragChanges = ({ dragStates, lastValues, }) => {
    const changes = [];
    for (const dragState of dragStates) {
        const value = lastValues.get(dragState.key);
        if (value === undefined) {
            continue;
        }
        if (dragState.target.propStatus.status === 'keyframed') {
            const startValue = (0, timeline_translate_utils_1.serializeTranslate)(dragState.startX, dragState.startY);
            if (value === startValue) {
                continue;
            }
            changes.push({
                type: 'keyframed',
                fileName: dragState.target.nodePath.absolutePath,
                nodePath: dragState.target.nodePath,
                fieldKey: translateFieldKey,
                sourceFrame: dragState.sourceFrame,
                value,
                schema: dragState.target.schema,
                clientId: dragState.target.clientId,
            });
            continue;
        }
        const stringifiedValue = JSON.stringify(value);
        const shouldSave = value !== dragState.target.propStatus.codeValue &&
            !(dragState.defaultValue === stringifiedValue &&
                dragState.target.propStatus.codeValue === undefined);
        if (!shouldSave) {
            continue;
        }
        changes.push({
            type: 'static',
            fileName: dragState.target.nodePath.absolutePath,
            nodePath: dragState.target.nodePath,
            fieldKey: translateFieldKey,
            value,
            defaultValue: dragState.defaultValue,
            schema: dragState.target.schema,
        });
    }
    return changes;
};
exports.getSelectedOutlineDragChanges = getSelectedOutlineDragChanges;
const getSelectedOutlineScaleEdgeInfo = (points, edge) => {
    const [tl, tr, br, bl] = points;
    const edgePoints = {
        top: { start: tl, end: tr, oppositeStart: bl, oppositeEnd: br },
        right: { start: tr, end: br, oppositeStart: tl, oppositeEnd: bl },
        bottom: { start: bl, end: br, oppositeStart: tl, oppositeEnd: tr },
        left: { start: tl, end: bl, oppositeStart: tr, oppositeEnd: br },
    }[edge];
    const edgeMidpoint = midpoint(edgePoints.start, edgePoints.end);
    const oppositeMidpoint = midpoint(edgePoints.oppositeStart, edgePoints.oppositeEnd);
    const outward = vectorBetween(oppositeMidpoint, edgeMidpoint);
    const length = vectorLength(outward);
    if (length < 0.001) {
        return null;
    }
    return {
        axis: edge === 'left' || edge === 'right' ? 'x' : 'y',
        cursor: edge === 'left' || edge === 'right' ? 'ew-resize' : 'ns-resize',
        end: edgePoints.end,
        extent: length,
        normal: { x: outward.x / length, y: outward.y / length },
        start: edgePoints.start,
    };
};
exports.getSelectedOutlineScaleEdgeInfo = getSelectedOutlineScaleEdgeInfo;
const getSelectedOutlineScaleDragStates = ({ dragTargets, getDragOverrides, }) => {
    return dragTargets.map((target) => {
        var _a;
        const dragOverrideValue = ((_a = getDragOverrides(target.nodePath)) !== null && _a !== void 0 ? _a : {})[scaleFieldKey];
        const effectiveValue = remotion_1.Internals.getEffectiveVisualModeValue({
            propStatus: target.propStatus,
            dragOverrideValue,
            defaultValue: target.fieldDefault,
            shouldResortToDefaultValueIfUndefined: true,
        });
        const [startX, startY, startZ] = no_react_1.NoReactInternals.parseScaleValue(effectiveValue);
        return {
            defaultValue: target.fieldDefault !== undefined
                ? JSON.stringify(target.fieldDefault)
                : null,
            key: remotion_1.Internals.makeSequencePropsSubscriptionKey(target.nodePath),
            startX,
            startY,
            startZ,
            target,
        };
    });
};
exports.getSelectedOutlineScaleDragStates = getSelectedOutlineScaleDragStates;
const getSelectedOutlineScaleDragValues = ({ axis, dragStates, scaleFactor, }) => {
    return new Map(dragStates.map((dragState) => {
        var _a, _b;
        const min = (_a = dragState.target.fieldSchema.min) !== null && _a !== void 0 ? _a : -Infinity;
        const max = (_b = dragState.target.fieldSchema.max) !== null && _b !== void 0 ? _b : Infinity;
        const baseX = dragState.startX;
        const baseY = dragState.startY;
        const newValue = (axis === 'x' ? baseX : baseY) * scaleFactor;
        const [x, y] = dragState.target.linked
            ? (0, TimelineScaleField_1.getLinkedScale)({
                axis,
                newValue,
                baseX,
                baseY,
                min,
                max,
            })
            : axis === 'x'
                ? [(0, selected_outline_geometry_1.clamp)(newValue, min, max), baseY]
                : [baseX, (0, selected_outline_geometry_1.clamp)(newValue, min, max)];
        return [
            dragState.key,
            no_react_1.NoReactInternals.serializeScaleValue([x, y, dragState.startZ]),
        ];
    }));
};
exports.getSelectedOutlineScaleDragValues = getSelectedOutlineScaleDragValues;
const getSelectedOutlineScaleDragChanges = ({ dragStates, lastValues, }) => {
    return dragStates.flatMap((dragState) => {
        const value = lastValues.get(dragState.key);
        if (value === undefined) {
            return [];
        }
        const stringifiedValue = JSON.stringify(value);
        const shouldSave = stringifiedValue !==
            JSON.stringify(dragState.target.propStatus.codeValue) &&
            !(dragState.defaultValue === stringifiedValue &&
                dragState.target.propStatus.codeValue === undefined);
        if (!shouldSave) {
            return [];
        }
        return [
            {
                fileName: dragState.target.nodePath.absolutePath,
                nodePath: dragState.target.nodePath,
                fieldKey: scaleFieldKey,
                value,
                defaultValue: dragState.defaultValue,
                schema: dragState.target.schema,
            },
        ];
    });
};
exports.getSelectedOutlineScaleDragChanges = getSelectedOutlineScaleDragChanges;
const getSelectedOutlineRotationDragStates = ({ dragTargets, getDragOverrides, timelinePosition, }) => {
    return dragTargets.map((target) => {
        var _a;
        const dragOverrideValue = ((_a = getDragOverrides(target.nodePath)) !== null && _a !== void 0 ? _a : {})[rotateFieldKey];
        const sourceFrame = timelinePosition - target.keyframeDisplayOffset;
        const effectiveValue = remotion_1.Internals.getEffectiveVisualModeValue({
            propStatus: target.propStatus,
            dragOverrideValue,
            defaultValue: target.fieldDefault,
            frame: sourceFrame,
            shouldResortToDefaultValueIfUndefined: true,
        });
        return {
            defaultValue: target.fieldDefault !== undefined
                ? JSON.stringify(target.fieldDefault)
                : null,
            key: remotion_1.Internals.makeSequencePropsSubscriptionKey(target.nodePath),
            sourceFrame,
            startDegrees: (0, timeline_rotation_utils_1.parseCssRotationToDegrees)(String(effectiveValue !== null && effectiveValue !== void 0 ? effectiveValue : '0deg')),
            target,
        };
    });
};
exports.getSelectedOutlineRotationDragStates = getSelectedOutlineRotationDragStates;
const getSelectedOutlineRotationDragValues = ({ dragStates, rotationDeltaDegrees, }) => {
    return new Map(dragStates.map((dragState) => [
        dragState.key,
        (0, timeline_rotation_utils_1.serializeCssRotation)(dragState.startDegrees + rotationDeltaDegrees),
    ]));
};
exports.getSelectedOutlineRotationDragValues = getSelectedOutlineRotationDragValues;
const getSelectedOutlineRotationDragChanges = ({ dragStates, lastValues, }) => {
    const changes = [];
    for (const dragState of dragStates) {
        const value = lastValues.get(dragState.key);
        if (value === undefined) {
            continue;
        }
        if (dragState.target.propStatus.status === 'keyframed') {
            const startValue = (0, timeline_rotation_utils_1.serializeCssRotation)(dragState.startDegrees);
            if (value === startValue) {
                continue;
            }
            changes.push({
                type: 'keyframed',
                fileName: dragState.target.nodePath.absolutePath,
                nodePath: dragState.target.nodePath,
                fieldKey: rotateFieldKey,
                sourceFrame: dragState.sourceFrame,
                value,
                schema: dragState.target.schema,
                clientId: dragState.target.clientId,
            });
            continue;
        }
        const stringifiedValue = JSON.stringify(value);
        const shouldSave = value !== dragState.target.propStatus.codeValue &&
            !(dragState.defaultValue === stringifiedValue &&
                dragState.target.propStatus.codeValue === undefined);
        if (!shouldSave) {
            continue;
        }
        changes.push({
            type: 'static',
            fileName: dragState.target.nodePath.absolutePath,
            nodePath: dragState.target.nodePath,
            fieldKey: rotateFieldKey,
            value,
            defaultValue: dragState.defaultValue,
            schema: dragState.target.schema,
        });
    }
    return changes;
};
exports.getSelectedOutlineRotationDragChanges = getSelectedOutlineRotationDragChanges;
const clearSelectedOutlineDragOverrides = ({ clearDragOverrides, dragStates, }) => {
    for (const dragState of dragStates) {
        clearDragOverrides(dragState.target.nodePath);
    }
};
const clearSelectedOutlineScaleDragOverrides = ({ clearDragOverrides, dragStates, }) => {
    for (const dragState of dragStates) {
        clearDragOverrides(dragState.target.nodePath);
    }
};
const clearSelectedOutlineRotationDragOverrides = ({ clearDragOverrides, dragStates, }) => {
    for (const dragState of dragStates) {
        clearDragOverrides(dragState.target.nodePath);
    }
};
const parseCssRotationToRadians = (value) => {
    const match = value
        .trim()
        .match(/^([+-]?(?:\d+\.?\d*|\.\d+))(deg|rad|turn|grad)$/);
    if (!match) {
        return null;
    }
    const number = Number(match[1]);
    if (!Number.isFinite(number)) {
        return null;
    }
    if (match[2] === 'rad') {
        return number;
    }
    if (match[2] === 'turn') {
        return number * Math.PI * 2;
    }
    if (match[2] === 'grad') {
        return (number / 400) * Math.PI * 2;
    }
    return (number / 180) * Math.PI;
};
const compensateTranslateForTransformOrigin = ({ startTranslate, deltaOrigin, rotate, scale, }) => {
    const cos = Math.cos(rotate);
    const sin = Math.sin(rotate);
    const matrixA = cos * scale[0];
    const matrixB = sin * scale[0];
    const matrixC = -sin * scale[1];
    const matrixD = cos * scale[1];
    const transformedDeltaX = matrixA * deltaOrigin[0] + matrixC * deltaOrigin[1];
    const transformedDeltaY = matrixB * deltaOrigin[0] + matrixD * deltaOrigin[1];
    const compensationX = deltaOrigin[0] - transformedDeltaX;
    const compensationY = deltaOrigin[1] - transformedDeltaY;
    return [startTranslate[0] - compensationX, startTranslate[1] - compensationY];
};
exports.compensateTranslateForTransformOrigin = compensateTranslateForTransformOrigin;
const uvsEqual = (left, right) => Math.abs(left[0] - right[0]) < 0.000001 &&
    Math.abs(left[1] - right[1]) < 0.000001;
const SelectedOutlineTransformOriginHandle = ({ outline, onDraggingChange, target }) => {
    var _a;
    const { setDragOverrides, clearDragOverrides, setPropStatuses } = (0, react_1.useContext)(remotion_1.Internals.VisualModeSettersContext);
    const transformOriginDrag = (_a = target === null || target === void 0 ? void 0 : target.transformOriginDrag) !== null && _a !== void 0 ? _a : null;
    const parsed = (0, react_1.useMemo)(() => transformOriginDrag === null
        ? null
        : (0, transform_origin_utils_1.parseTransformOrigin)(transformOriginDrag.originValue), [transformOriginDrag]);
    const uv = (0, react_1.useMemo)(() => {
        if (parsed === null || outline.dimensions === null) {
            return null;
        }
        return (0, transform_origin_utils_1.parsedTransformOriginToUv)({
            parsed,
            width: outline.dimensions.width,
            height: outline.dimensions.height,
        });
    }, [outline.dimensions, parsed]);
    const position = (0, react_1.useMemo)(() => (uv === null ? null : (0, selected_outline_uv_1.getUvHandlePosition)(outline.points, uv)), [outline.points, uv]);
    const onPointerDown = react_1.default.useCallback((event) => {
        if (event.button !== 0 ||
            transformOriginDrag === null ||
            parsed === null ||
            uv === null ||
            outline.dimensions === null) {
            return;
        }
        event.preventDefault();
        event.stopPropagation();
        const svg = event.currentTarget.ownerSVGElement;
        if (svg === null) {
            return;
        }
        const rotation = parseCssRotationToRadians(transformOriginDrag.rotateValue);
        if (rotation === null) {
            return;
        }
        const { dimensions } = outline;
        if (dimensions === null) {
            return;
        }
        const [scaleX, scaleY] = no_react_1.NoReactInternals.parseScaleValue(transformOriginDrag.scaleValue);
        const startTranslate = (0, timeline_translate_utils_1.parseTranslate)(transformOriginDrag.translateValue);
        const svgRect = svg.getBoundingClientRect();
        const defaultOrigin = transformOriginDrag.originDefault !== undefined
            ? JSON.stringify(transformOriginDrag.originDefault)
            : null;
        const defaultTranslate = transformOriginDrag.translateDefault !== undefined
            ? JSON.stringify(transformOriginDrag.translateDefault)
            : null;
        let last = null;
        onDraggingChange(true);
        (0, ForceSpecificCursor_1.forceSpecificCursor)('crosshair');
        const updateFromPointerEvent = (pointerEvent) => {
            const point = {
                x: pointerEvent.clientX - svgRect.left,
                y: pointerEvent.clientY - svgRect.top,
            };
            const nextUv = (0, selected_outline_uv_1.getUvCoordinateForPoint)(outline.points, point);
            const deltaOrigin = [
                (nextUv[0] - uv[0]) * dimensions.width,
                (nextUv[1] - uv[1]) * dimensions.height,
            ];
            const [nextTranslateX, nextTranslateY] = (0, exports.compensateTranslateForTransformOrigin)({
                startTranslate,
                deltaOrigin,
                rotate: rotation,
                scale: [scaleX, scaleY],
            });
            const origin = (0, transform_origin_utils_1.serializeTransformOrigin)({
                uv: nextUv,
                z: parsed.z,
            });
            const translate = (0, timeline_translate_utils_1.serializeTranslate)(nextTranslateX, nextTranslateY);
            last = { uv: nextUv, origin, translate };
            setDragOverrides(transformOriginDrag.nodePath, transformOriginFieldKey, transformOriginDrag.originPropStatus.status === 'keyframed'
                ? remotion_1.Internals.makeKeyframedDragOverride({
                    status: transformOriginDrag.originPropStatus,
                    frame: transformOriginDrag.sourceFrame,
                    value: origin,
                })
                : remotion_1.Internals.makeStaticDragOverride(origin));
            setDragOverrides(transformOriginDrag.nodePath, translateFieldKey, transformOriginDrag.translatePropStatus.status === 'keyframed'
                ? remotion_1.Internals.makeKeyframedDragOverride({
                    status: transformOriginDrag.translatePropStatus,
                    frame: transformOriginDrag.sourceFrame,
                    value: translate,
                })
                : remotion_1.Internals.makeStaticDragOverride(translate));
        };
        updateFromPointerEvent(event);
        const onPointerMove = (moveEvent) => {
            moveEvent.preventDefault();
            updateFromPointerEvent(moveEvent);
        };
        const onPointerUp = () => {
            window.removeEventListener('pointermove', onPointerMove);
            window.removeEventListener('pointerup', onPointerUp);
            window.removeEventListener('pointercancel', onPointerUp);
            (0, ForceSpecificCursor_1.stopForcingSpecificCursor)();
            onDraggingChange(false);
            if (last === null || uvsEqual(last.uv, uv)) {
                clearDragOverrides(transformOriginDrag.nodePath);
                return;
            }
            const originChanged = last.origin !== transformOriginDrag.originValue;
            const translateChanged = last.translate !== transformOriginDrag.translateValue;
            if (!originChanged && !translateChanged) {
                clearDragOverrides(transformOriginDrag.nodePath);
                return;
            }
            const shouldSaveAsKeyframes = transformOriginDrag.originPropStatus.status === 'keyframed' ||
                transformOriginDrag.translatePropStatus.status === 'keyframed';
            const promise = shouldSaveAsKeyframes
                ? (0, call_add_keyframe_1.callAddKeyframes)({
                    sequenceKeyframes: [
                        originChanged
                            ? {
                                fileName: transformOriginDrag.nodePath.absolutePath,
                                nodePath: transformOriginDrag.nodePath,
                                fieldKey: transformOriginFieldKey,
                                sourceFrame: transformOriginDrag.sourceFrame,
                                value: last.origin,
                                schema: transformOriginDrag.schema,
                            }
                            : null,
                        translateChanged
                            ? {
                                fileName: transformOriginDrag.nodePath.absolutePath,
                                nodePath: transformOriginDrag.nodePath,
                                fieldKey: translateFieldKey,
                                sourceFrame: transformOriginDrag.sourceFrame,
                                value: last.translate,
                                schema: transformOriginDrag.schema,
                            }
                            : null,
                    ].filter(no_react_1.NoReactInternals.truthy),
                    effectKeyframes: [],
                    setPropStatuses,
                    clientId: transformOriginDrag.clientId,
                })
                : (0, save_sequence_prop_1.saveSequenceProps)({
                    changes: [
                        originChanged
                            ? {
                                fileName: transformOriginDrag.nodePath.absolutePath,
                                nodePath: transformOriginDrag.nodePath,
                                fieldKey: transformOriginFieldKey,
                                value: last.origin,
                                defaultValue: defaultOrigin,
                                schema: transformOriginDrag.schema,
                            }
                            : null,
                        translateChanged
                            ? {
                                fileName: transformOriginDrag.nodePath.absolutePath,
                                nodePath: transformOriginDrag.nodePath,
                                fieldKey: translateFieldKey,
                                value: last.translate,
                                defaultValue: defaultTranslate,
                                schema: transformOriginDrag.schema,
                            }
                            : null,
                    ].filter(no_react_1.NoReactInternals.truthy),
                    setPropStatuses,
                    clientId: transformOriginDrag.clientId,
                    undoLabel: 'Move transform origin',
                    redoLabel: 'Move transform origin back',
                });
            promise
                .catch((err) => {
                (0, NotificationCenter_1.showNotification)(`Could not save transform origin: ${err instanceof Error ? err.message : String(err)}`, 4000);
            })
                .finally(() => {
                clearDragOverrides(transformOriginDrag.nodePath);
            });
        };
        window.addEventListener('pointermove', onPointerMove);
        window.addEventListener('pointerup', onPointerUp);
        window.addEventListener('pointercancel', onPointerUp);
    }, [
        clearDragOverrides,
        onDraggingChange,
        outline,
        parsed,
        setDragOverrides,
        setPropStatuses,
        transformOriginDrag,
        uv,
    ]);
    if (transformOriginDrag === null ||
        parsed === null ||
        uv === null ||
        position === null) {
        return null;
    }
    return (jsx_runtime_1.jsxs("g", { pointerEvents: "all", cursor: "crosshair", onPointerDown: onPointerDown, "aria-hidden": "true", children: [
            jsx_runtime_1.jsx("line", { x1: position.x - 8, y1: position.y, x2: position.x + 8, y2: position.y, stroke: "white", strokeWidth: 4, vectorEffect: "non-scaling-stroke" }), jsx_runtime_1.jsx("line", { x1: position.x, y1: position.y - 8, x2: position.x, y2: position.y + 8, stroke: "white", strokeWidth: 4, vectorEffect: "non-scaling-stroke" }), jsx_runtime_1.jsx("circle", { cx: position.x, cy: position.y, r: 4, fill: "white", stroke: colors_1.BLUE, strokeWidth: 2, vectorEffect: "non-scaling-stroke" }), jsx_runtime_1.jsx("line", { x1: position.x - 8, y1: position.y, x2: position.x + 8, y2: position.y, stroke: colors_1.BLUE, strokeWidth: 2, vectorEffect: "non-scaling-stroke" }), jsx_runtime_1.jsx("line", { x1: position.x, y1: position.y - 8, x2: position.x, y2: position.y + 8, stroke: colors_1.BLUE, strokeWidth: 2, vectorEffect: "non-scaling-stroke" })
        ] }));
};
const SelectedOutlinePolygon = ({ allDragTargets, contextMenuValues, dragging, hovered, onContextMenuOpen, outline, onDraggingChange, onHoverChange, onSelect, scale, target, }) => {
    var _a, _b, _c, _d;
    const { getDragOverrides } = (0, react_1.useContext)(remotion_1.Internals.VisualModeDragOverridesContext);
    const { setPropStatuses, setDragOverrides, clearDragOverrides } = (0, react_1.useContext)(remotion_1.Internals.VisualModeSettersContext);
    const timelinePosition = remotion_1.Internals.Timeline.useTimelinePosition();
    const timelinePositionRef = (0, react_1.useRef)(timelinePosition);
    timelinePositionRef.current = timelinePosition;
    const polygonRef = (0, react_1.useRef)(null);
    const points = (0, react_1.useMemo)(() => outline.points.map(pointToString).join(' '), [outline.points]);
    const drag = (_a = target === null || target === void 0 ? void 0 : target.drag) !== null && _a !== void 0 ? _a : null;
    const selected = (_b = target === null || target === void 0 ? void 0 : target.selected) !== null && _b !== void 0 ? _b : false;
    const containsSelection = (_c = target === null || target === void 0 ? void 0 : target.containsSelection) !== null && _c !== void 0 ? _c : false;
    const effectDrop = (_d = target === null || target === void 0 ? void 0 : target.effectDrop) !== null && _d !== void 0 ? _d : null;
    const [effectDropHovered, setEffectDropHovered] = (0, react_1.useState)(false);
    const visible = containsSelection || hovered;
    const onPointerDown = react_1.default.useCallback((event) => {
        if (event.button !== 0 || target === undefined) {
            return;
        }
        event.preventDefault();
        event.stopPropagation();
        const interaction = (0, exports.getOutlineSelectionInteraction)(event);
        const shouldUpdateSelection = !selected || interaction.shiftKey || interaction.toggleKey;
        if (shouldUpdateSelection) {
            onSelect(target.selection, interaction);
        }
        if (drag === null || interaction.shiftKey || interaction.toggleKey) {
            return;
        }
        onDraggingChange(true);
        const startPointerX = event.clientX;
        const startPointerY = event.clientY;
        const dragStates = getSelectedOutlineDragStates({
            dragTargets: selected ? allDragTargets : [drag],
            getDragOverrides,
            timelinePosition: timelinePositionRef.current,
        });
        let lastValues = new Map();
        let currentPointerX = startPointerX;
        let currentPointerY = startPointerY;
        let axisLocked = false;
        const updateDragOverrides = () => {
            const dragDelta = (0, exports.applySelectedOutlineDragAxisLock)({
                deltaX: (currentPointerX - startPointerX) / scale,
                deltaY: (currentPointerY - startPointerY) / scale,
                axisLocked,
            });
            lastValues = (0, exports.getSelectedOutlineDragValues)({
                dragStates,
                deltaX: dragDelta.deltaX,
                deltaY: dragDelta.deltaY,
            });
            for (const dragState of dragStates) {
                const value = lastValues.get(dragState.key);
                if (value === undefined) {
                    throw new Error('Expected drag value to be available');
                }
                if (dragState.target.propStatus.status === 'keyframed') {
                    setDragOverrides(dragState.target.nodePath, translateFieldKey, remotion_1.Internals.makeKeyframedDragOverride({
                        status: dragState.target.propStatus,
                        frame: dragState.sourceFrame,
                        value,
                    }));
                }
                else {
                    setDragOverrides(dragState.target.nodePath, translateFieldKey, remotion_1.Internals.makeStaticDragOverride(value));
                }
            }
        };
        const onPointerMove = (moveEvent) => {
            moveEvent.preventDefault();
            currentPointerX = moveEvent.clientX;
            currentPointerY = moveEvent.clientY;
            axisLocked = moveEvent.shiftKey;
            updateDragOverrides();
        };
        const onKeyChange = (keyEvent) => {
            if (keyEvent.key !== 'Shift') {
                return;
            }
            const nextAxisLocked = keyEvent.type === 'keydown';
            if (nextAxisLocked === axisLocked) {
                return;
            }
            axisLocked = nextAxisLocked;
            updateDragOverrides();
        };
        const onPointerUp = () => {
            window.removeEventListener('pointermove', onPointerMove);
            window.removeEventListener('pointerup', onPointerUp);
            window.removeEventListener('pointercancel', onPointerUp);
            window.removeEventListener('keydown', onKeyChange);
            window.removeEventListener('keyup', onKeyChange);
            onDraggingChange(false);
            const changes = (0, exports.getSelectedOutlineDragChanges)({
                dragStates,
                lastValues,
            });
            if (changes.length === 0) {
                clearSelectedOutlineDragOverrides({ clearDragOverrides, dragStates });
                return;
            }
            const staticChanges = changes.filter((change) => change.type === 'static');
            const keyframedChanges = changes.filter((change) => change.type === 'keyframed');
            Promise.all([
                staticChanges.length > 0
                    ? (0, save_sequence_prop_1.saveSequenceProps)({
                        changes: staticChanges,
                        setPropStatuses,
                        clientId: drag.clientId,
                        undoLabel: changes.length > 1
                            ? 'Move selected sequences'
                            : 'Move sequence',
                        redoLabel: changes.length > 1
                            ? 'Move selected sequences back'
                            : 'Move sequence back',
                    })
                    : Promise.resolve(),
                ...keyframedChanges.map((change) => (0, call_add_keyframe_1.callAddSequenceKeyframe)({
                    fileName: change.fileName,
                    nodePath: change.nodePath,
                    fieldKey: change.fieldKey,
                    sourceFrame: change.sourceFrame,
                    value: change.value,
                    schema: change.schema,
                    setPropStatuses,
                    clientId: change.clientId,
                })),
            ])
                .catch((err) => {
                (0, NotificationCenter_1.showNotification)(`Could not save sequence props: ${err instanceof Error ? err.message : String(err)}`, 4000);
            })
                .finally(() => {
                clearSelectedOutlineDragOverrides({ clearDragOverrides, dragStates });
            });
        };
        window.addEventListener('pointermove', onPointerMove);
        window.addEventListener('pointerup', onPointerUp);
        window.addEventListener('pointercancel', onPointerUp);
        window.addEventListener('keydown', onKeyChange);
        window.addEventListener('keyup', onKeyChange);
    }, [
        allDragTargets,
        clearDragOverrides,
        drag,
        getDragOverrides,
        onDraggingChange,
        onSelect,
        scale,
        selected,
        setPropStatuses,
        setDragOverrides,
        target,
    ]);
    const onEffectDragOver = react_1.default.useCallback((event) => {
        if (effectDrop === null || !(0, effect_drag_and_drop_1.hasEffectDragType)(event.dataTransfer)) {
            return;
        }
        event.preventDefault();
        event.stopPropagation();
        event.dataTransfer.dropEffect = 'copy';
        setEffectDropHovered(true);
    }, [effectDrop]);
    const onEffectDragLeave = react_1.default.useCallback((event) => {
        if (event.currentTarget.contains(event.relatedTarget)) {
            return;
        }
        setEffectDropHovered(false);
    }, []);
    const onEffectDrop = react_1.default.useCallback(async (event) => {
        if (effectDrop === null || !(0, effect_drag_and_drop_1.hasEffectDragType)(event.dataTransfer)) {
            return;
        }
        const dragData = (0, effect_drag_and_drop_1.getEffectDragData)(event.dataTransfer);
        if (!dragData) {
            if ((0, effect_drag_and_drop_1.hasExplicitEffectDragType)(event.dataTransfer)) {
                event.preventDefault();
                event.stopPropagation();
                setEffectDropHovered(false);
                (0, NotificationCenter_1.showNotification)('Could not read effect drag data', 3000);
            }
            return;
        }
        event.preventDefault();
        event.stopPropagation();
        setEffectDropHovered(false);
        await (0, effect_drag_and_drop_1.addEffectFromDragData)({
            dragData,
            fileName: effectDrop.fileName,
            nodePath: effectDrop.nodePath,
            clientId: effectDrop.clientId,
        });
    }, [effectDrop]);
    return (jsx_runtime_1.jsxs(jsx_runtime_1.Fragment, { children: [
            jsx_runtime_1.jsx("polygon", { ref: polygonRef, points: points, fill: effectDropHovered ? 'rgba(0, 155, 255, 0.12)' : 'transparent', stroke: colors_1.BLUE, strokeOpacity: visible || effectDropHovered ? 1 : 0, strokeWidth: 2, vectorEffect: "non-scaling-stroke", pointerEvents: target === undefined ? undefined : 'all', onPointerEnter: () => {
                    if (!dragging) {
                        onHoverChange(outline.key);
                    }
                }, onPointerLeave: () => {
                    if (!dragging) {
                        onHoverChange(null);
                    }
                }, onPointerDown: onPointerDown, onDragOver: effectDrop === null ? undefined : onEffectDragOver, onDragLeave: effectDrop === null ? undefined : onEffectDragLeave, onDrop: effectDrop === null ? undefined : onEffectDrop }), jsx_runtime_1.jsx(ContextMenu_1.ContextMenuForTarget, { triggerRef: polygonRef, values: [...contextMenuValues], onOpen: onContextMenuOpen })
        ] }));
};
const SelectedOutlineScaleEdgeLine = ({ allScaleDragTargets, contextMenuValues, dragging, edge, outline, onDraggingChange, onContextMenuOpen, onHoverChange, onSelect, target, }) => {
    var _a, _b;
    const { getDragOverrides } = (0, react_1.useContext)(remotion_1.Internals.VisualModeDragOverridesContext);
    const { setPropStatuses, setDragOverrides, clearDragOverrides } = (0, react_1.useContext)(remotion_1.Internals.VisualModeSettersContext);
    const scaleDrag = (_a = target === null || target === void 0 ? void 0 : target.scaleDrag) !== null && _a !== void 0 ? _a : null;
    const selected = (_b = target === null || target === void 0 ? void 0 : target.selected) !== null && _b !== void 0 ? _b : false;
    const lineRef = (0, react_1.useRef)(null);
    const edgeInfo = (0, react_1.useMemo)(() => (0, exports.getSelectedOutlineScaleEdgeInfo)(outline.points, edge), [edge, outline.points]);
    const onPointerDown = react_1.default.useCallback((event) => {
        if (event.button !== 0 || scaleDrag === null || edgeInfo === null) {
            return;
        }
        event.preventDefault();
        event.stopPropagation();
        const interaction = (0, exports.getOutlineSelectionInteraction)(event);
        const shouldUpdateSelection = !selected || interaction.shiftKey || interaction.toggleKey;
        if (shouldUpdateSelection && target !== undefined) {
            onSelect(target.selection, interaction);
        }
        if (interaction.shiftKey || interaction.toggleKey) {
            return;
        }
        onDraggingChange(true);
        const startPointer = { x: event.clientX, y: event.clientY };
        const dragStates = (0, exports.getSelectedOutlineScaleDragStates)({
            dragTargets: selected ? allScaleDragTargets : [scaleDrag],
            getDragOverrides,
        });
        let lastValues = new Map();
        const onPointerMove = (moveEvent) => {
            moveEvent.preventDefault();
            const delta = {
                x: moveEvent.clientX - startPointer.x,
                y: moveEvent.clientY - startPointer.y,
            };
            const projectedDelta = dot(delta, edgeInfo.normal);
            const scaleFactor = Math.max(0.001, 1 + projectedDelta / edgeInfo.extent);
            lastValues = (0, exports.getSelectedOutlineScaleDragValues)({
                dragStates,
                axis: edgeInfo.axis,
                scaleFactor,
            });
            for (const dragState of dragStates) {
                const value = lastValues.get(dragState.key);
                if (value === undefined) {
                    throw new Error('Expected scale drag value to be available');
                }
                setDragOverrides(dragState.target.nodePath, scaleFieldKey, remotion_1.Internals.makeStaticDragOverride(value));
            }
        };
        const onPointerUp = () => {
            window.removeEventListener('pointermove', onPointerMove);
            window.removeEventListener('pointerup', onPointerUp);
            window.removeEventListener('pointercancel', onPointerUp);
            onDraggingChange(false);
            const changes = (0, exports.getSelectedOutlineScaleDragChanges)({
                dragStates,
                lastValues,
            });
            if (changes.length === 0) {
                clearSelectedOutlineScaleDragOverrides({
                    clearDragOverrides,
                    dragStates,
                });
                return;
            }
            (0, save_sequence_prop_1.saveSequenceProps)({
                changes,
                setPropStatuses,
                clientId: scaleDrag.clientId,
                undoLabel: changes.length > 1 ? 'Scale selected sequences' : 'Scale sequence',
                redoLabel: changes.length > 1
                    ? 'Scale selected sequences back'
                    : 'Scale sequence back',
            })
                .catch((err) => {
                (0, NotificationCenter_1.showNotification)(`Could not save sequence props: ${err instanceof Error ? err.message : String(err)}`, 4000);
            })
                .finally(() => {
                clearSelectedOutlineScaleDragOverrides({
                    clearDragOverrides,
                    dragStates,
                });
            });
        };
        window.addEventListener('pointermove', onPointerMove);
        window.addEventListener('pointerup', onPointerUp);
        window.addEventListener('pointercancel', onPointerUp);
    }, [
        allScaleDragTargets,
        clearDragOverrides,
        edgeInfo,
        getDragOverrides,
        onDraggingChange,
        onSelect,
        scaleDrag,
        selected,
        setPropStatuses,
        setDragOverrides,
        target,
    ]);
    if (scaleDrag === null || edgeInfo === null) {
        return null;
    }
    return (jsx_runtime_1.jsxs(jsx_runtime_1.Fragment, { children: [
            jsx_runtime_1.jsx("line", { ref: lineRef, x1: edgeInfo.start.x, y1: edgeInfo.start.y, x2: edgeInfo.end.x, y2: edgeInfo.end.y, stroke: "transparent", strokeWidth: 12, vectorEffect: "non-scaling-stroke", pointerEvents: "stroke", cursor: edgeInfo.cursor, onPointerEnter: () => {
                    if (!dragging) {
                        onHoverChange(outline.key);
                    }
                }, onPointerLeave: () => {
                    if (!dragging) {
                        onHoverChange(null);
                    }
                }, onPointerDown: onPointerDown }), jsx_runtime_1.jsx(ContextMenu_1.ContextMenuForTarget, { triggerRef: lineRef, values: [...contextMenuValues], onOpen: onContextMenuOpen })
        ] }));
};
const svgPointToClientPoint = (point, rect) => {
    return {
        x: point.x + rect.left,
        y: point.y + rect.top,
    };
};
const SelectedOutlineRotationCornerHandle = ({ allRotationDragTargets, contextMenuValues, corner, dragging, outline, onDraggingChange, onContextMenuOpen, onHoverChange, onSelect, target, }) => {
    var _a, _b;
    const { getDragOverrides } = (0, react_1.useContext)(remotion_1.Internals.VisualModeDragOverridesContext);
    const { setPropStatuses, setDragOverrides, clearDragOverrides } = (0, react_1.useContext)(remotion_1.Internals.VisualModeSettersContext);
    const timelinePosition = remotion_1.Internals.Timeline.useTimelinePosition();
    const timelinePositionRef = (0, react_1.useRef)(timelinePosition);
    timelinePositionRef.current = timelinePosition;
    const rotationDrag = (_a = target === null || target === void 0 ? void 0 : target.rotationDrag) !== null && _a !== void 0 ? _a : null;
    const selected = (_b = target === null || target === void 0 ? void 0 : target.selected) !== null && _b !== void 0 ? _b : false;
    const circleRef = (0, react_1.useRef)(null);
    const cornerInfo = (0, react_1.useMemo)(() => (0, exports.getSelectedOutlineRotationCornerInfo)(outline.points, corner), [corner, outline.points]);
    const onPointerDown = react_1.default.useCallback((event) => {
        if (event.button !== 0 || rotationDrag === null) {
            return;
        }
        event.preventDefault();
        event.stopPropagation();
        const svg = event.currentTarget.ownerSVGElement;
        if (svg === null) {
            return;
        }
        const interaction = (0, exports.getOutlineSelectionInteraction)(event);
        const shouldUpdateSelection = !selected || interaction.shiftKey || interaction.toggleKey;
        if (shouldUpdateSelection && target !== undefined) {
            onSelect(target.selection, interaction);
        }
        if (interaction.shiftKey || interaction.toggleKey) {
            return;
        }
        onDraggingChange(true);
        (0, ForceSpecificCursor_1.forceSpecificCursor)(cornerInfo.cursor);
        const svgRect = svg.getBoundingClientRect();
        const center = svgPointToClientPoint(cornerInfo.center, svgRect);
        const dragStates = (0, exports.getSelectedOutlineRotationDragStates)({
            dragTargets: selected ? allRotationDragTargets : [rotationDrag],
            getDragOverrides,
            timelinePosition: timelinePositionRef.current,
        });
        let previousAngle = getAngleDegrees(center, {
            x: event.clientX,
            y: event.clientY,
        });
        let accumulatedDelta = 0;
        let lastValues = new Map();
        const onPointerMove = (moveEvent) => {
            moveEvent.preventDefault();
            const nextAngle = getAngleDegrees(center, {
                x: moveEvent.clientX,
                y: moveEvent.clientY,
            });
            accumulatedDelta += (0, exports.getSelectedOutlineRotationDeltaDegrees)({
                from: previousAngle,
                to: nextAngle,
            });
            previousAngle = nextAngle;
            lastValues = (0, exports.getSelectedOutlineRotationDragValues)({
                dragStates,
                rotationDeltaDegrees: accumulatedDelta,
            });
            (0, ForceSpecificCursor_1.forceSpecificCursor)(getRotationCursor(cornerInfo.cursorDegrees + accumulatedDelta));
            for (const dragState of dragStates) {
                const value = lastValues.get(dragState.key);
                if (value === undefined) {
                    throw new Error('Expected rotation drag value to be available');
                }
                if (dragState.target.propStatus.status === 'keyframed') {
                    setDragOverrides(dragState.target.nodePath, rotateFieldKey, remotion_1.Internals.makeKeyframedDragOverride({
                        status: dragState.target.propStatus,
                        frame: dragState.sourceFrame,
                        value,
                    }));
                }
                else {
                    setDragOverrides(dragState.target.nodePath, rotateFieldKey, remotion_1.Internals.makeStaticDragOverride(value));
                }
            }
        };
        const onPointerUp = () => {
            window.removeEventListener('pointermove', onPointerMove);
            window.removeEventListener('pointerup', onPointerUp);
            window.removeEventListener('pointercancel', onPointerUp);
            (0, ForceSpecificCursor_1.stopForcingSpecificCursor)();
            onDraggingChange(false);
            const changes = (0, exports.getSelectedOutlineRotationDragChanges)({
                dragStates,
                lastValues,
            });
            if (changes.length === 0) {
                clearSelectedOutlineRotationDragOverrides({
                    clearDragOverrides,
                    dragStates,
                });
                return;
            }
            const staticChanges = changes.filter((change) => change.type === 'static');
            const keyframedChanges = changes.filter((change) => change.type === 'keyframed');
            Promise.all([
                staticChanges.length > 0
                    ? (0, save_sequence_prop_1.saveSequenceProps)({
                        changes: staticChanges,
                        setPropStatuses,
                        clientId: rotationDrag.clientId,
                        undoLabel: changes.length > 1
                            ? 'Rotate selected sequences'
                            : 'Rotate sequence',
                        redoLabel: changes.length > 1
                            ? 'Rotate selected sequences back'
                            : 'Rotate sequence back',
                    })
                    : Promise.resolve(),
                ...keyframedChanges.map((change) => (0, call_add_keyframe_1.callAddSequenceKeyframe)({
                    fileName: change.fileName,
                    nodePath: change.nodePath,
                    fieldKey: change.fieldKey,
                    sourceFrame: change.sourceFrame,
                    value: change.value,
                    schema: change.schema,
                    setPropStatuses,
                    clientId: change.clientId,
                })),
            ])
                .catch((err) => {
                (0, NotificationCenter_1.showNotification)(`Could not save sequence props: ${err instanceof Error ? err.message : String(err)}`, 4000);
            })
                .finally(() => {
                clearSelectedOutlineRotationDragOverrides({
                    clearDragOverrides,
                    dragStates,
                });
            });
        };
        window.addEventListener('pointermove', onPointerMove);
        window.addEventListener('pointerup', onPointerUp);
        window.addEventListener('pointercancel', onPointerUp);
    }, [
        allRotationDragTargets,
        clearDragOverrides,
        cornerInfo,
        getDragOverrides,
        onDraggingChange,
        onSelect,
        rotationDrag,
        selected,
        setPropStatuses,
        setDragOverrides,
        target,
    ]);
    if (rotationDrag === null) {
        return null;
    }
    return (jsx_runtime_1.jsxs(jsx_runtime_1.Fragment, { children: [
            jsx_runtime_1.jsx("circle", { ref: circleRef, cx: cornerInfo.point.x, cy: cornerInfo.point.y, r: 12, fill: "transparent", stroke: "transparent", vectorEffect: "non-scaling-stroke", pointerEvents: "all", cursor: cornerInfo.cursor, onPointerEnter: () => {
                    if (!dragging) {
                        onHoverChange(outline.key);
                    }
                }, onPointerLeave: () => {
                    if (!dragging) {
                        onHoverChange(null);
                    }
                }, onPointerDown: onPointerDown }), jsx_runtime_1.jsx(ContextMenu_1.ContextMenuForTarget, { triggerRef: circleRef, values: [...contextMenuValues], onOpen: onContextMenuOpen })
        ] }));
};
const SelectedOutlineElement = ({ allDragTargets, allRotationDragTargets, allScaleDragTargets, dragging, hovered, outline, onDraggingChange, onHoverChange, onSelect, scale, target, }) => {
    const { previewServerState } = (0, react_1.useContext)(client_id_1.StudioServerConnectionCtx);
    const updateResolvedStackTrace = (0, react_1.useContext)(remotion_1.Internals.SequenceStackTracesUpdateContext);
    const onContextMenuOpen = react_1.default.useCallback(async () => {
        if (target === undefined || previewServerState.type !== 'connected') {
            return false;
        }
        if (!target.selected) {
            onSelect(target.selection, { shiftKey: false, toggleKey: false });
        }
        const stack = target.sequence.getStack();
        let originalLocation = null;
        if (stack) {
            try {
                originalLocation = await (0, get_stack_1.getOriginalLocationFromStack)(stack, 'sequence');
            }
            catch (err) {
                (0, NotificationCenter_1.showNotification)(err.message, 2000);
            }
        }
        if (stack) {
            updateResolvedStackTrace(stack, originalLocation);
        }
        const fileLocation = (0, format_file_location_1.formatFileLocation)({
            location: originalLocation,
            root: window.remotion_cwd,
        });
        const editorName = window.remotion_editorName;
        return [
            editorName
                ? {
                    type: 'item',
                    id: 'show-outline-in-editor',
                    keyHint: null,
                    label: `Show in ${editorName}`,
                    leftItem: null,
                    disabled: !originalLocation,
                    onClick: () => {
                        if (!originalLocation) {
                            return;
                        }
                        (0, open_in_editor_1.openOriginalPositionInEditor)(originalLocation).catch((err) => {
                            (0, NotificationCenter_1.showNotification)(err.message, 2000);
                        });
                    },
                    quickSwitcherLabel: null,
                    subMenu: null,
                    value: 'show-outline-in-editor',
                }
                : null,
            {
                type: 'item',
                id: 'copy-outline-file-location',
                keyHint: null,
                label: 'Copy file location',
                leftItem: null,
                disabled: !fileLocation,
                onClick: () => {
                    if (!fileLocation) {
                        return;
                    }
                    navigator.clipboard
                        .writeText(fileLocation)
                        .then(() => {
                        (0, NotificationCenter_1.showNotification)('Copied file location to clipboard', 1000);
                    })
                        .catch((err) => {
                        (0, NotificationCenter_1.showNotification)(`Could not copy to clipboard: ${err.message}`, 1000);
                    });
                },
                quickSwitcherLabel: null,
                subMenu: null,
                value: 'copy-outline-file-location',
            },
        ].filter(no_react_1.NoReactInternals.truthy);
    }, [onSelect, previewServerState.type, target, updateResolvedStackTrace]);
    return (jsx_runtime_1.jsxs(jsx_runtime_1.Fragment, { children: [
            jsx_runtime_1.jsx(SelectedOutlinePolygon, { allDragTargets: allDragTargets, contextMenuValues: emptyContextMenuValues, dragging: dragging, hovered: hovered, outline: outline, onContextMenuOpen: onContextMenuOpen, onDraggingChange: onDraggingChange, onHoverChange: onHoverChange, onSelect: onSelect, scale: scale, target: target }), (target === null || target === void 0 ? void 0 : target.containsSelection) || hovered
                ? ['top', 'right', 'bottom', 'left'].map((edge) => (jsx_runtime_1.jsx(SelectedOutlineScaleEdgeLine, { allScaleDragTargets: allScaleDragTargets, contextMenuValues: emptyContextMenuValues, dragging: dragging, edge: edge, outline: outline, onContextMenuOpen: onContextMenuOpen, onDraggingChange: onDraggingChange, onHoverChange: onHoverChange, onSelect: onSelect, target: target }, edge)))
                : null, (target === null || target === void 0 ? void 0 : target.containsSelection) || hovered
                ? ['top-left', 'top-right', 'bottom-right', 'bottom-left'].map((corner) => (jsx_runtime_1.jsx(SelectedOutlineRotationCornerHandle, { allRotationDragTargets: allRotationDragTargets, contextMenuValues: emptyContextMenuValues, corner: corner, dragging: dragging, outline: outline, onContextMenuOpen: onContextMenuOpen, onDraggingChange: onDraggingChange, onHoverChange: onHoverChange, onSelect: onSelect, target: target }, corner)))
                : null, jsx_runtime_1.jsx(SelectedOutlineTransformOriginHandle, { outline: outline, onDraggingChange: onDraggingChange, target: target })
        ] }));
};
const SelectedOutlineOverlay = ({ scale }) => {
    const { selectedItems, selectItem } = (0, TimelineSelection_1.useTimelineSelection)();
    const { sequences } = (0, react_1.useContext)(remotion_1.Internals.SequenceManager);
    const { propStatuses } = (0, react_1.useContext)(remotion_1.Internals.VisualModePropStatusesContext);
    const { previewServerState } = (0, react_1.useContext)(client_id_1.StudioServerConnectionCtx);
    const { overrideIdToNodePathMappings } = (0, react_1.useContext)(remotion_1.Internals.OverrideIdsToNodePathsGettersContext);
    const { getDragOverrides, getEffectDragOverrides } = (0, react_1.useContext)(remotion_1.Internals.VisualModeDragOverridesContext);
    const { getScaleLockState } = (0, react_1.useContext)(scale_lock_1.ScaleLockContext);
    const { editorShowOutlines } = (0, react_1.useContext)(editor_outlines_1.EditorShowOutlinesContext);
    const timelinePosition = remotion_1.Internals.Timeline.useTimelinePosition();
    const [outlines, setOutlines] = (0, react_1.useState)([]);
    const [hoveredOutlineKey, setHoveredOutlineKey] = (0, react_1.useState)(null);
    const [draggingOutline, setDraggingOutline] = (0, react_1.useState)(false);
    const overlayRef = (0, react_1.useRef)(null);
    const onDraggingChange = react_1.default.useCallback((dragging) => {
        setDraggingOutline(dragging);
        if (dragging) {
            setHoveredOutlineKey(null);
        }
    }, []);
    const outlineTargets = (0, react_1.useMemo)(() => {
        if (!editorShowOutlines) {
            return [];
        }
        const selectedSequenceKeys = (0, exports.getSelectedSequenceKeys)(selectedItems);
        const sequenceKeysContainingSelection = getSequenceKeysContainingSelection(selectedItems);
        const selectedEffectsBySequenceKey = (0, exports.getSelectedEffectFieldsBySequenceKey)(selectedItems);
        const selectedTransformOriginInfo = getSelectedTransformOriginInfo(selectedItems);
        const clientId = previewServerState.type === 'connected'
            ? previewServerState.clientId
            : null;
        return (0, exports.getSequencesWithSelectableOutlines)({
            sequences,
            overrideIdsToNodePaths: overrideIdToNodePathMappings,
        }).map(({ key, keyframeDisplayOffset, nodePathInfo, sequence }) => {
            var _a;
            var _b, _c, _d, _e, _f, _g, _h, _j;
            if (sequence.refForOutline === null) {
                throw new Error('Expected sequence to have a ref for outline');
            }
            const selected = selectedSequenceKeys.has(key);
            const containsSelection = sequenceKeysContainingSelection.has(key);
            const nodePath = nodePathInfo.sequenceSubscriptionKey;
            const { controls } = sequence;
            const fieldSchema = controls === null || controls === void 0 ? void 0 : controls.schema[translateFieldKey];
            const nodePropStatuses = remotion_1.Internals.getPropStatusesCtx(propStatuses, nodePath);
            const propStatus = nodePropStatuses === null || nodePropStatuses === void 0 ? void 0 : nodePropStatuses[translateFieldKey];
            const scaleFieldSchema = controls === null || controls === void 0 ? void 0 : controls.schema[scaleFieldKey];
            const scalePropStatus = nodePropStatuses === null || nodePropStatuses === void 0 ? void 0 : nodePropStatuses[scaleFieldKey];
            const rotationFieldSchema = controls === null || controls === void 0 ? void 0 : controls.schema[rotateFieldKey];
            const rotationPropStatus = (_a = remotion_1.Internals.getPropStatusesCtx(propStatuses, nodePath)) === null || _a === void 0 ? void 0 : _a[rotateFieldKey];
            const transformOriginFieldSchema = controls === null || controls === void 0 ? void 0 : controls.schema[transformOriginFieldKey];
            const transformOriginPropStatus = nodePropStatuses === null || nodePropStatuses === void 0 ? void 0 : nodePropStatuses[transformOriginFieldKey];
            const canDragStatus = (propStatus === null || propStatus === void 0 ? void 0 : propStatus.status) === 'static' ||
                ((propStatus === null || propStatus === void 0 ? void 0 : propStatus.status) === 'keyframed' &&
                    propStatus.interpolationFunction === 'interpolate');
            const canRotationDragStatus = (rotationPropStatus === null || rotationPropStatus === void 0 ? void 0 : rotationPropStatus.status) === 'static' ||
                ((rotationPropStatus === null || rotationPropStatus === void 0 ? void 0 : rotationPropStatus.status) === 'keyframed' &&
                    rotationPropStatus.interpolationFunction === 'interpolate');
            const canDrag = previewServerState.type === 'connected' &&
                controls !== null &&
                (fieldSchema === null || fieldSchema === void 0 ? void 0 : fieldSchema.type) === 'translate' &&
                canDragStatus;
            const canScaleDrag = previewServerState.type === 'connected' &&
                controls !== null &&
                (scaleFieldSchema === null || scaleFieldSchema === void 0 ? void 0 : scaleFieldSchema.type) === 'scale' &&
                (scalePropStatus === null || scalePropStatus === void 0 ? void 0 : scalePropStatus.status) === 'static';
            const canRotationDrag = previewServerState.type === 'connected' &&
                controls !== null &&
                (rotationFieldSchema === null || rotationFieldSchema === void 0 ? void 0 : rotationFieldSchema.type) === 'rotation-css' &&
                canRotationDragStatus;
            const selectedForTransformOrigin = (selectedTransformOriginInfo === null || selectedTransformOriginInfo === void 0 ? void 0 : selectedTransformOriginInfo.sequenceKey) === key;
            const transformOriginSourceFrame = (selectedTransformOriginInfo === null || selectedTransformOriginInfo === void 0 ? void 0 : selectedTransformOriginInfo.displayFrame) === null ||
                (selectedTransformOriginInfo === null || selectedTransformOriginInfo === void 0 ? void 0 : selectedTransformOriginInfo.displayFrame) === undefined
                ? timelinePosition - keyframeDisplayOffset
                : selectedTransformOriginInfo.displayFrame - keyframeDisplayOffset;
            const canTransformOriginStatus = (transformOriginPropStatus === null || transformOriginPropStatus === void 0 ? void 0 : transformOriginPropStatus.status) === 'static' ||
                ((transformOriginPropStatus === null || transformOriginPropStatus === void 0 ? void 0 : transformOriginPropStatus.status) === 'keyframed' &&
                    transformOriginPropStatus.interpolationFunction === 'interpolate');
            const canTransformOriginTranslateStatus = (propStatus === null || propStatus === void 0 ? void 0 : propStatus.status) === 'static' ||
                ((propStatus === null || propStatus === void 0 ? void 0 : propStatus.status) === 'keyframed' &&
                    propStatus.interpolationFunction === 'interpolate');
            const canTransformOriginDrag = previewServerState.type === 'connected' &&
                selectedForTransformOrigin &&
                controls !== null &&
                (transformOriginFieldSchema === null || transformOriginFieldSchema === void 0 ? void 0 : transformOriginFieldSchema.type) === 'transform-origin' &&
                (fieldSchema === null || fieldSchema === void 0 ? void 0 : fieldSchema.type) === 'translate' &&
                canTransformOriginStatus &&
                canTransformOriginTranslateStatus;
            const canDropEffect = previewServerState.type === 'connected' &&
                (controls === null || controls === void 0 ? void 0 : controls.supportsEffects) === true;
            return {
                key,
                containsSelection,
                effectDrop: canDropEffect
                    ? {
                        clientId: previewServerState.clientId,
                        fileName: nodePath.absolutePath,
                        nodePath,
                    }
                    : null,
                nodePathInfo,
                ref: sequence.refForOutline,
                selected,
                selection: { type: 'sequence', nodePathInfo },
                sequence,
                drag: canDrag
                    ? {
                        propStatus,
                        clientId: previewServerState.clientId,
                        fieldDefault: fieldSchema.default,
                        keyframeDisplayOffset,
                        nodePath,
                        schema: controls.schema,
                    }
                    : null,
                scaleDrag: canScaleDrag
                    ? {
                        propStatus: scalePropStatus,
                        clientId: previewServerState.clientId,
                        fieldDefault: scaleFieldSchema.default,
                        fieldSchema: scaleFieldSchema,
                        linked: getScaleLockState({
                            nodePath,
                            fieldKey: scaleFieldKey,
                            defaultValue: (() => {
                                var _a;
                                const dragOverrideValue = ((_a = getDragOverrides(nodePath)) !== null && _a !== void 0 ? _a : {})[scaleFieldKey];
                                const effectiveValue = remotion_1.Internals.getEffectiveVisualModeValue({
                                    propStatus: scalePropStatus,
                                    dragOverrideValue,
                                    defaultValue: scaleFieldSchema.default,
                                    shouldResortToDefaultValueIfUndefined: true,
                                });
                                const [x, y] = no_react_1.NoReactInternals.parseScaleValue(effectiveValue);
                                return x === y;
                            })(),
                        }),
                        nodePath,
                        schema: controls.schema,
                    }
                    : null,
                rotationDrag: canRotationDrag
                    ? {
                        propStatus: rotationPropStatus,
                        clientId: previewServerState.clientId,
                        fieldDefault: rotationFieldSchema.default,
                        fieldSchema: rotationFieldSchema,
                        keyframeDisplayOffset,
                        nodePath,
                        schema: controls.schema,
                    }
                    : null,
                transformOriginDrag: canTransformOriginDrag
                    ? {
                        clientId: previewServerState.clientId,
                        keyframeDisplayOffset,
                        nodePath,
                        originDefault: transformOriginFieldSchema.default,
                        originPropStatus: transformOriginPropStatus,
                        originValue: String((_c = remotion_1.Internals.getEffectiveVisualModeValue({
                            propStatus: transformOriginPropStatus,
                            dragOverrideValue: ((_b = getDragOverrides(nodePath)) !== null && _b !== void 0 ? _b : {})[transformOriginFieldKey],
                            defaultValue: transformOriginFieldSchema.default,
                            frame: transformOriginSourceFrame,
                            shouldResortToDefaultValueIfUndefined: true,
                        })) !== null && _c !== void 0 ? _c : transformOriginFieldSchema.default),
                        rotateValue: String((rotationPropStatus === null || rotationPropStatus === void 0 ? void 0 : rotationPropStatus.status) === 'static' ||
                            (rotationPropStatus === null || rotationPropStatus === void 0 ? void 0 : rotationPropStatus.status) === 'keyframed'
                            ? ((_e = remotion_1.Internals.getEffectiveVisualModeValue({
                                propStatus: rotationPropStatus,
                                dragOverrideValue: ((_d = getDragOverrides(nodePath)) !== null && _d !== void 0 ? _d : {})[rotateFieldKey],
                                defaultValue: (rotationFieldSchema === null || rotationFieldSchema === void 0 ? void 0 : rotationFieldSchema.type) === 'rotation-css'
                                    ? rotationFieldSchema.default
                                    : '0deg',
                                frame: transformOriginSourceFrame,
                                shouldResortToDefaultValueIfUndefined: true,
                            })) !== null && _e !== void 0 ? _e : '0deg')
                            : '0deg'),
                        scaleValue: (scalePropStatus === null || scalePropStatus === void 0 ? void 0 : scalePropStatus.status) === 'static' ||
                            (scalePropStatus === null || scalePropStatus === void 0 ? void 0 : scalePropStatus.status) === 'keyframed'
                            ? String((_g = remotion_1.Internals.getEffectiveVisualModeValue({
                                propStatus: scalePropStatus,
                                dragOverrideValue: ((_f = getDragOverrides(nodePath)) !== null && _f !== void 0 ? _f : {})[scaleFieldKey],
                                defaultValue: (scaleFieldSchema === null || scaleFieldSchema === void 0 ? void 0 : scaleFieldSchema.type) === 'scale'
                                    ? scaleFieldSchema.default
                                    : 1,
                                frame: transformOriginSourceFrame,
                                shouldResortToDefaultValueIfUndefined: true,
                            })) !== null && _g !== void 0 ? _g : 1)
                            : '1',
                        schema: controls.schema,
                        sourceFrame: transformOriginSourceFrame,
                        translateDefault: fieldSchema.default,
                        translatePropStatus: propStatus,
                        translateValue: String((_j = remotion_1.Internals.getEffectiveVisualModeValue({
                            propStatus,
                            dragOverrideValue: ((_h = getDragOverrides(nodePath)) !== null && _h !== void 0 ? _h : {})[translateFieldKey],
                            defaultValue: fieldSchema.default,
                            frame: transformOriginSourceFrame,
                            shouldResortToDefaultValueIfUndefined: true,
                        })) !== null && _j !== void 0 ? _j : fieldSchema.default),
                    }
                    : null,
                uvHandles: containsSelection
                    ? (0, selected_outline_uv_1.getSelectedUvHandles)({
                        propStatuses,
                        clientId,
                        getEffectDragOverrides,
                        nodePath,
                        selectedEffects: selectedEffectsBySequenceKey.get(key),
                        sequence,
                        sourceFrame: timelinePosition - keyframeDisplayOffset,
                    })
                    : [],
            };
        });
    }, [
        propStatuses,
        getDragOverrides,
        getEffectDragOverrides,
        getScaleLockState,
        editorShowOutlines,
        overrideIdToNodePathMappings,
        previewServerState,
        selectedItems,
        sequences,
        timelinePosition,
    ]);
    (0, react_1.useEffect)(() => {
        if (hoveredOutlineKey !== null &&
            !outlineTargets.some((target) => target.key === hoveredOutlineKey)) {
            setHoveredOutlineKey(null);
        }
    }, [hoveredOutlineKey, outlineTargets]);
    const targetsByKey = (0, react_1.useMemo)(() => {
        return new Map(outlineTargets.map((target) => [target.key, target]));
    }, [outlineTargets]);
    const allDragTargets = (0, react_1.useMemo)(() => {
        return outlineTargets.flatMap((target) => target.selected && target.drag !== null ? [target.drag] : []);
    }, [outlineTargets]);
    const allScaleDragTargets = (0, react_1.useMemo)(() => {
        return outlineTargets.flatMap((target) => target.selected && target.scaleDrag !== null ? [target.scaleDrag] : []);
    }, [outlineTargets]);
    const allRotationDragTargets = (0, react_1.useMemo)(() => {
        return outlineTargets.flatMap((target) => target.selected && target.rotationDrag !== null
            ? [target.rotationDrag]
            : []);
    }, [outlineTargets]);
    (0, react_1.useEffect)(() => {
        if (outlineTargets.length === 0) {
            setOutlines((prevOutlines) => prevOutlines.length === 0 ? prevOutlines : []);
            return;
        }
        let animationFrame = null;
        const updateOutlines = () => {
            if (overlayRef.current) {
                const nextOutlines = measureOutlines(overlayRef.current, outlineTargets);
                setOutlines((prevOutlines) => outlinesAreEqual(prevOutlines, nextOutlines)
                    ? prevOutlines
                    : nextOutlines);
            }
            animationFrame = requestAnimationFrame(updateOutlines);
        };
        updateOutlines();
        return () => {
            if (animationFrame !== null) {
                cancelAnimationFrame(animationFrame);
            }
        };
    }, [outlineTargets]);
    if (outlineTargets.length === 0) {
        return null;
    }
    return (jsx_runtime_1.jsxs("svg", { ref: overlayRef, style: outlineContainer, width: "100%", height: "100%", "aria-hidden": "true", children: [outlines.map((outline) => (jsx_runtime_1.jsx(SelectedOutlineElement, { allDragTargets: allDragTargets, allRotationDragTargets: allRotationDragTargets, allScaleDragTargets: allScaleDragTargets, dragging: draggingOutline, hovered: hoveredOutlineKey === outline.key, outline: outline, onDraggingChange: onDraggingChange, onHoverChange: setHoveredOutlineKey, onSelect: selectItem, scale: scale, target: targetsByKey.get(outline.key) }, outline.key))), outlines.map((outline) => (jsx_runtime_1.jsx(SelectedOutlineUvControls_1.SelectedOutlineUvHandleConnectionLayer, { outline: outline, target: targetsByKey.get(outline.key) }, `${outline.key}-uv-connection-lines`))), outlines.map((outline) => (jsx_runtime_1.jsx(SelectedOutlineUvControls_1.SelectedOutlineUvHandleCircleLayer, { onDraggingChange: onDraggingChange, outline: outline, target: targetsByKey.get(outline.key) }, `${outline.key}-uv-handles`)))] }));
};
exports.SelectedOutlineOverlay = SelectedOutlineOverlay;
